pkg/standalone-utils/test/MidasWrapping.test.ts

```diff
@@ -0,0 +1,775 @@
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+import { BigNumber, Contract, ContractReceipt } from 'ethers';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers';
+
+import Token from '@balancer-labs/v2-helpers/src/models/tokens/Token';
+import TokenList from '@balancer-labs/v2-helpers/src/models/tokens/TokenList';
+import StablePool from '@balancer-labs/v2-helpers/src/models/pools/stable/StablePool';
+
+import { SwapKind, StablePoolEncoder } from '@balancer-labs/balancer-js';
+import * as expectEvent from '@balancer-labs/v2-helpers/src/test/expectEvent';
+import { expectTransferEvent } from '@balancer-labs/v2-helpers/src/test/expectTransfer';
+import { deploy, deployedAt } from '@balancer-labs/v2-helpers/src/contract';
+import { actionId } from '@balancer-labs/v2-helpers/src/models/misc/actions';
+import { MAX_INT256, MAX_UINT256, ZERO_ADDRESS } from '@balancer-labs/v2-helpers/src/constants';
+import { BigNumberish, fp, bn } from '@balancer-labs/v2-helpers/src/numbers';
+import Vault from '@balancer-labs/v2-helpers/src/models/vault/Vault';
+import { Account } from '@balancer-labs/v2-helpers/src/models/types/types';
+import TypesConverter from '@balancer-labs/v2-helpers/src/models/types/TypesConverter';
+import { Dictionary } from 'lodash';
+import {
+  expectChainedReferenceContents,
+  setChainedReferenceContents,
+  toChainedReference,
+} from './helpers/chainedReferences';
+
+describe('MidasWrapping', function () {
+  let DAI: Token, cDAI: Token;
+  let senderUser: SignerWithAddress, recipientUser: SignerWithAddress, admin: SignerWithAddress;
+  let vault: Vault;
+  let relayer: Contract, relayerLibrary: Contract;
+
+  before('setup signer', async () => {
+    [, admin, senderUser, recipientUser] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('deploy Vault', async () => {
+    // const initialExchangeRate: BigNumber;
+    vault = await Vault.create({ admin });
+
+    const wrappedTokenDecimals = 8;
+    const mainTokenDecimals = 18;
+    // cToken exchangeRate is an unsigned integer scaled by 1 * 10 ^(18-8+underlyingDecimals)
+    // const initialExchangeRate = bn(1e28);
+
+    DAI = await deploy('v2-solidity-utils/TestToken', { args: ['DAI', 'DAI', mainTokenDecimals] });
+
+    cDAI = await deploy('MockCToken', {
+      args: ['cDAI', 'cDAI', wrappedTokenDecimals, DAI.address, bn(1e18)],
+    });
+  });
+
+  sharedBeforeEach('mint tokens to senderUser', async () => {
+    await DAI.mint(senderUser.address, fp(100));
+    await DAI.connect(senderUser).approve(vault.address, fp(100));
+    await DAI.mint(cDAI.address, fp(10000));
+
+    await cDAI.mintCTokens(senderUser.address, fp(100));
+    await cDAI.connect(senderUser).approve(cDAI.address, fp(150));
```

Comment:
Approve the Vault here so we don't need to do it later. The self-approval is meaningless anyway.
```suggestion
    await cDAI.connect(senderUser).approve(vault.address, fp(150));
```